include/nbl/system/IThreadHandler.h

```diff
@@ -0,0 +1,103 @@
+#ifndef __NBL_I_THREAD_HANDLER_H_INCLUDED__
+#define __NBL_I_THREAD_HANDLER_H_INCLUDED__
+
+#include <mutex>
+#include <condition_variable>
+#include <thread>
+
+namespace nbl {
+namespace system
+{
+
+// Usage:
+/*
+* class MyThreadHandler : public IThreadHandler<SomeInternalStateType> { .... };
+* 
+* MyThreadHandler handler;
+* std::thread thread(&MyThreadHandler::thread, &handler);
+* //...
+* //... communicate with the thread using your methods (see note at the end of this section)), thread will sleep until wakeupPredicate() returns true
+* //...
+* handler.terminate(thread);
+* After this handler can be safely destroyed.
+* Every method playing around with object's state shared with the thread must begin with line: `auto raii_handler = createRAIIDisptachHandler();`!
+*/
+template <typename InternalStateType = int>
+class IThreadHandler
+{
+protected:
+    using mutex_t = std::mutex;
+    using cvar_t = std::condition_variable;
+    using lock_t = std::unique_lock<mutex_t>;
+
+    using internal_state_t = InternalStateType;
+
+    struct raii_dispatch_handler_t
+    {
+        raii_dispatch_handler_t(lock_t&& _lk, cvar_t& _cv) : lk(std::move(lk)), cv(_cv) {}
+        ~raii_dispatch_handler_t()
+        {
+            lk.unlock();
+            cv.notify_one();
+        }
+
+    private:
+        lock_t lk;
+        cvar_t& cv;
+    };
+
+    inline lock_t createLock() { return lock_t{ m_mutex }; }
+    inline raii_dispatch_handler_t createRAIIDispatchHandler() { return raii_dispatch_handler_t(createLock(), m_cvar); }
+
+    virtual internal_state_t init() = 0;
+    virtual bool wakeupPredicate() const { return m_quit; }
+    virtual bool continuePredicate() const { return !m_quit; }
+
+    // lock is locked at the beginning of this function and must be locked at the exit
+    virtual void work(lock_t& lock, internal_state_t& state) = 0;
+
+    virtual void exit(internal_state_t& state) {}
```

Comment:
use CRTP over virtualism

`internal_state_t` could come from `typename CRTP::internal_state_t` I guess?

Reply:
maybe it doesnt even need to be a Recurring Template Parameter, but just a parameter

Reply:
it is a template parameter, i dont even know what you mean with this virtualism and CRTP here XD

Reply:
actually there's probably no point of keeping track of `internal_state_t` if the Handler to Thread ratio is 1:1 because it could just be member variables that the derived class has access to at all times.

Reply:
no, because it must be initialized by the thread, thats the whole point of its existence

Reply:
> 
> 
> no, because it must be initialized by the thread, thats the whole point of its existence

sure and you can do that by keeping it as a member if there's only 1 thread per handler

Reply:
> 
> 
> it is a template parameter, i dont even know what you mean with this virtualism and CRTP here XD

prefer static polymorphism to dynamic/runtime polymorphism

Reply:
>sure and you can do that by keeping it as a member if there's only 1 thread per handler

no, then it's initialized in thread in which the handler object is created

Reply:
idk what's the problem with init method